core/shared/src/main/scala/zio/stm/TQueue.scala

```diff
@@ -54,9 +54,75 @@ final class TQueue[A] private (val capacity: Int, ref: TRef[ScalaQueue[A]]) {
       .map(_.splitAt(max))
       .flatMap(split => ref.set(split._2) *> STM.succeed(split._1))
       .map(_.toList)
+
+  /**
+   * View the next element in the queue without removing it. retries if the queue is empty.
+   */
+  def peek: STM[Nothing, A] =
+    ref.get.flatMap(
+      _.headOption match {
+        case Some(a) => STM.succeed(a)
+        case None    => STM.retry
+      }
+    )
+
+  /**
+   * View the last element inserted into the queue. retries if the queue is empty
+   */
+  def back: STM[Nothing, A] =
+    ref.get.flatMap(
+      _.lastOption match {
+        case Some(a) => STM.succeed(a)
+        case None    => STM.retry
+      }
+    )
+
+  /**
+   * Checks if the Queue is empty
+   */
+  def isEmpty: STM[Nothing, Boolean] = ref.get.map(_.isEmpty)
+
+  /**
+   * Checks if the Queue is at capacity
+   */
+  def isFull: STM[Nothing, Boolean] = ref.get.map(_.size == capacity)
+
+  /**
+   * Drops elements from the queue while they dont satisfy the predicate until it does and takes it.
+   */
+  def seek(f: A => Boolean): STM[Nothing, A] = {
+    @annotation.tailrec
+    def go(q: ScalaQueue[A]): STM[Nothing, A] =
+      q.dequeueOption match {
+        case Some((a, as)) =>
+          if (f(a)) ref.set(as) *> STM.succeed(a)
+          else go(as)
+        case None => STM.retry
+      }
+
+    ref.get.flatMap(go)
+  }
+
+  /**
+   * Splits the queue into a prefix/suffix pair according to a predicate.
+   */
+  def span(f: A => Boolean): STM[Nothing, (TQueue[A], TQueue[A])] =
```

Comment:
`partition`?

Reply:
This is the one I'm not too sure about. In general, anything in `STM` that returns another structure is probably not that useful. STM is designed for modifying structures, not producing new structures. I think I'd delete this one but keep the others.